hbase-server/src/main/java/org/apache/hadoop/hbase/master/root/RootTable.java

```diff
@@ -0,0 +1,148 @@
+
+  static final String COMPACT_MIN_KEY = "hbase.root.table.region.compact.min";
+
+  private static final int DEFAULT_COMPACT_MIN = 4;
+
+  static final String ROLL_PERIOD_MS_KEY = "hbase.root.table.region.walroll.period.ms";
+
+  private static final long DEFAULT_ROLL_PERIOD_MS = TimeUnit.MINUTES.toMillis(15);
+
+  static final TableName TABLE_NAME = TableName.valueOf("master:root");
+
+  private static final TableDescriptor TABLE_DESC = TableDescriptorBuilder.newBuilder(TABLE_NAME)
+    .setColumnFamily(ColumnFamilyDescriptorBuilder.newBuilder(HConstants.CATALOG_FAMILY)
+      .setMaxVersions(HConstants.DEFAULT_HBASE_META_VERSIONS).setInMemory(true)
+      .setBlocksize(HConstants.DEFAULT_HBASE_META_BLOCK_SIZE).setBloomFilterType(BloomType.ROWCOL)
+      .setDataBlockEncoding(DataBlockEncoding.ROW_INDEX_V1).build())
+    .build();
+
+  private final Server server;
+
+  private final DirScanPool cleanerPool;
+
+  private LocalRegion region;
+
+  public RootTable(Server server, DirScanPool cleanerPool) {
+    this.server = server;
+    this.cleanerPool = cleanerPool;
+  }
+
+  public void initialize() throws IOException {
+    LOG.info("Initializing root table...");
+    LocalRegionParams params = new LocalRegionParams().server(server).regionDirName(ROOT_TABLE_DIR)
+      .tableDescriptor(TABLE_DESC);
+    Configuration conf = server.getConfiguration();
+    long flushSize = conf.getLong(FLUSH_SIZE_KEY, DEFAULT_FLUSH_SIZE);
+    long flushPerChanges = conf.getLong(FLUSH_PER_CHANGES_KEY, DEFAULT_FLUSH_PER_CHANGES);
+    long flushIntervalMs = conf.getLong(FLUSH_INTERVAL_MS_KEY, DEFAULT_FLUSH_INTERVAL_MS);
+    int compactMin = conf.getInt(COMPACT_MIN_KEY, DEFAULT_COMPACT_MIN);
+    params.flushSize(flushSize).flushPerChanges(flushPerChanges).flushIntervalMs(flushIntervalMs)
+      .compactMin(compactMin);
+    int maxWals = conf.getInt(MAX_WALS_KEY, DEFAULT_MAX_WALS);
+    params.maxWals(maxWals);
+    if (conf.get(USE_HSYNC_KEY) != null) {
+      params.useHsync(conf.getBoolean(USE_HSYNC_KEY, false));
+    }
+    params.ringBufferSlotCount(conf.getInt(RING_BUFFER_SLOT_COUNT, DEFAULT_RING_BUFFER_SLOT_COUNT));
+    long rollPeriodMs = conf.getLong(ROLL_PERIOD_MS_KEY, DEFAULT_ROLL_PERIOD_MS);
+    params.rollPeriodMs(rollPeriodMs)
+      .archivedWalSuffix(MasterProcedureUtil.ARCHIVED_PROC_WAL_SUFFIX)
+      .hfileCleanerPlugins(HFILECLEANER_PLUGINS).cleanerPool(cleanerPool);
+    region = LocalRegion.create(params);
+  }
+
+  public void update(Put put) throws IOException {
+    region.update(r -> r.put(put));
+  }
+
+  public void delete(Delete delete) throws IOException {
+    region.update(r -> r.delete(delete));
+  }
+
+  public RegionScanner getScanner(Scan scan) throws IOException {
+    return region.getScanner(scan);
+  }
+
+  public void close(boolean abort) {
+    LOG.info("Closing root table, isAbort={}", abort);
+    if (region != null) {
+      region.close(abort);
+    }
+  }
+}
```

Comment:
Yes. I've posted on jira, we can introduce a single region, maybe called 'master-local-store', to store all the master local data, such as the root table and procedures. If we think this is a better way, we should be hurry to finish the work before 2.3.0 is out, otherwise we need to deal with the migration from the procedure store region again...